patroni/validator.py

```diff
@@ -0,0 +1,230 @@
+        logging.basicConfig(level=logging.DEBUG, format='%(levelname)-8s %(message)s')
+        for key in self._schema:
+            optional = False
+            if isinstance(self._schema, dict):
+                subschema = self._schema[key]
+            elif isinstance(self._schema, list):
+                subschema = self._schema[0]
+                for key, value in enumerate(data):
+                    self.validator(key, value, subschema)
+                continue
+            if not isinstance(key, str):
+                if isinstance(key, Optional):
+                    key = key.name
+                    optional = True
+                if isinstance(key, Or):
+                    self.validate_orcase(key, subschema, data)
+                    continue
+            if key not in data and not optional:
+                logger.warning("%s is not defined", self.get_fullpath(key))
+            else:
+                self.validator(key, data[key], subschema)
+
+    def validator(self, key, data, schema, quiet=False):
+        if not isinstance(data, type(schema)):
+            if issubclass(type(schema), type):
+                if isinstance(data, schema):
+                    return True
+                else:
+                    if not quiet:
+                        logger.warning("%s ('%s') didn't pass validation", self.get_fullpath(key), data)
+            elif callable(schema):
+                try:
+                    schema(data)
+                    return True
+                except Exception as e:
+                    if not quiet:
+                        logger.warning("%s ('%s') didn't pass validation: %s", self.get_fullpath(key), data, e)
+            elif isinstance(schema, Or):
+                counter = 0
+                for v in schema.args:
+                    counter += (1 if self.validator(key, data, v, quiet=True) else 0)
+                if counter < 1:
+                    logger.warning("%s ('%s') didn't pass validation", self.get_fullpath(key), data)
+                else:
+                    return True
+
+            else:
+                if not quiet:
+                    logger.warning("%s type is not %s", self.get_fullpath(key), _get_type_name(type(schema)))
+        elif isinstance(data, dict):
+            Schema(schema, path=self.get_fullpath(key))(data)
+        elif isinstance(data, list):
+            Schema(schema, path=self.get_fullpath(key))(data)
+
+    def validate_orcase(self, key, schema, data):
+        counter = 0
+        for a in key.args:
+            if a in data:
+                counter += 1
+                subschema = schema._schema[a]
+                self.validator(a, data[a], subschema)
+        if counter < 1:
+            logger.warning("neither of %s is defined", ", ".join([self.get_fullpath(a) for a in key.args]))
+
+    def get_fullpath(self, path):
+        return self._path + ("." if self._path else "") + str(path)
+
+
+def _get_type_name(python_type):
```

Comment:
Looks like it is unused.

Reply:
https://github.com/zalando/patroni/pull/1314/files#diff-05020fb6b44e1d780d1bb03b5b0e257eR143